sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala

```diff
@@ -193,6 +194,87 @@ case class MapPartitionsExec(
   }
 }
 
+/**
+ * Similar with [[MapPartitionsExec]] and
+ * [[org.apache.spark.sql.execution.r.MapPartitionsRWrapper]] but serializes and deserializes
+ * input/output in Arrow format.
+ *
+ * This is somewhat similar with [[org.apache.spark.sql.execution.python.ArrowEvalPythonExec]]
+ */
+case class MapPartitionsInRWithArrowExec(
+    func: Array[Byte],
+    packageNames: Array[Byte],
+    broadcastVars: Array[Broadcast[Object]],
+    inputSchema: StructType,
+    output: Seq[Attribute],
+    child: SparkPlan) extends UnaryExecNode {
+
+  private val batchSize = conf.arrowMaxRecordsPerBatch
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    child.execute().mapPartitionsInternal { iter =>
+      // Hyukjin: seems to convert from Arrow batches to Arrow batches directly here
+      // causes segmentation fault intermittently. For instance, when child is directly
+      // originated from Arrow batches. Input projection was added to avoid it based on
+      // my speculation (and it does not happen anymore). It is potentially an issue in Arrow
+      // optimization in my humble opinion.
```

Comment:
I don't have an explicit evidence for this.

Reply:
This only happens when the input batchs are pruned by https://github.com/apache/spark/pull/23810. The root cause seems when the input columns are less than the schema set to Arrow writer.

I am going to work around by the same way. In that way, it's guaranteed that the input columns are not being pruned comparing to Arrow writer.

Reply:
Seems this happens only when the input is pruned but Arrow writer expects full schema. I will follow the same approach of https://github.com/apache/spark/pull/23810 for `gapply()`.
Therefore, the input will be guaranteed to not be pruned.